src/QsCompiler/CompilationManager/EditorSupport/SyntaxUtils.cs

```diff
@@ -0,0 +1,328 @@
+                _ => null,
+            };
+
+            return child ?? expression;
+
+            TypedExpression? Binary(TypedExpression e1, TypedExpression e2) =>
+                FindExpressionByPosition(e1, position) ?? FindExpressionByPosition(e2, position);
+
+            TypedExpression? Many(IEnumerable<TypedExpression> es) =>
+                es.Select(e => FindExpressionByPosition(e, position)).FirstOrDefault(e => !(e is null));
+        }
+
+        private static IEnumerable<TypedExpression> ExpressionsInStatement(QsStatement statement)
+        {
+            return statement.Statement switch
+            {
+                QsStatementKind.QsExpressionStatement expression => new[] { expression.Item },
+                QsStatementKind.QsReturnStatement @return => new[] { @return.Item },
+                QsStatementKind.QsFailStatement fail => new[] { fail.Item },
+                QsStatementKind.QsVariableDeclaration declaration => new[] { declaration.Item.Rhs },
+                QsStatementKind.QsValueUpdate update => new[] { update.Item.Rhs },
+                QsStatementKind.QsConditionalStatement cond => cond.Item.ConditionalBlocks.Select(CondBlockExpression),
+                QsStatementKind.QsForStatement @for => new[] { @for.Item.IterationValues },
+                QsStatementKind.QsWhileStatement @while => new[] { @while.Item.Condition },
+                QsStatementKind.QsRepeatStatement repeat => new[] { repeat.Item.SuccessCondition },
+                QsStatementKind.QsQubitScope qubit => InitializerExpressions(qubit.Item.Binding.Rhs),
+                _ => Enumerable.Empty<TypedExpression>(),
+            };
+
+            TypedExpression CondBlockExpression(Tuple<TypedExpression, QsPositionedBlock> b)
+            {
+                if (statement.Location.IsNull || b.Item2.Location.IsNull)
+                {
+                    return b.Item1;
+                }
+
+                var offset = b.Item2.Location.Item.Offset - statement.Location.Item.Offset;
+                return new ExpressionOffsetTransformation(offset).OnTypedExpression(b.Item1);
+            }
+
+            static IEnumerable<TypedExpression> InitializerExpressions(ResolvedInitializer i) => i.Resolution switch
+            {
+                QsInitializerKind.QubitRegisterAllocation r => new[] { r.Item },
+                QsInitializerKind.QubitTupleAllocation t => t.Item.SelectMany(InitializerExpressions),
+                _ => Enumerable.Empty<TypedExpression>(),
+            };
+        }
+
+        private static IEnumerable<LocalVariableDeclaration<string>> DeclarationsInExpressionByPosition(
+            TypedExpression expression, Position position)
+        {
+            if (expression.Range.IsNull || !expression.Range.Item.Contains(position))
+            {
+                return Enumerable.Empty<LocalVariableDeclaration<string>>();
+            }
+
+            return expression.Expression switch
```

Comment:
It feels like there is an opportunity to unify the expression matching in DeclarationsInExpressionByPosition and in FindExpressionByPosition by defining a private helper that takes suitable functions as arguments.

Reply:
I made the F# type extensions Fold and ExtractAll accessible to C# and used those here.

Reply:
I like that - that's much cleaner.